internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
+	// Review: the spec handles state and state changes a bit differently
```

Comment:
Yes -- we introduced this change as what we thought would be the simplest way to mitigate reentrancy attacks. The basic invariants are:
- `AcquireState` and `Release|UpdateRelease` calls must be balanced and only one state lock may be held at a time
- `Send*` calls cannot be made while the state lock is held
The runtime should enforce these properties dynamically (see, e.g., `_checkStateLock`). We would like to enforce as much as possible statically, but couldn't see a good way to do this without introducing a lot of complexity at the API level and making the actor code highly non-Go-idiomatic.

Reply:
I have too many questions on this one..
Do we have any docs describing this "reentrancy attack"? I want to see what it is you are trying to protect against.

Reply:
I sent you some references out of band.

Reply:
The spec is not actually locking, and is not clear how this is `StateHandle` state is shared between reentrant calls to actor. 

I created the `ActorStateHandler` type and added the methods, left some `Review:` comments on it too.
We can take the discussion over there since there are quite a few distinct concerns.

Reply:
> The spec is not actually locking

In the context of a single method invocation, the effect of the runtime checks is to lock the state -- we don't allow concurrent calls, so we don't use inter-thread synchronization. But to the actor code, the state appears locked: it can't be re-acquired, it must be released, and no `Send` call can be made while it's held.

Reply:
> not clear how this is StateHandle state is shared between reentrant calls to actor

Reentrant calls should never need to share a state handle -- in fact, this can't happen since method calls should be the only mechanism for reentrancy, but this is prohibited if a state handle is held.

Reply:
lets talk offline